src/code/m_npc.c

```diff
@@ -0,0 +1,3975 @@
+        npcMax[0]++;
+    }
+}
+
+void mNpc_SubNowNpcMax(u8* npcMax) {
+    if (npcMax[0] > ANIMAL_NUM_MIN) {
+        npcMax[0]--;
+    }
+}
+
+void mNpc_ClearAnimalPersonalID(AnmPersonalID_c* pid) {
+    pid->npcId = 0;
+    pid->nameId = 0xFF;
+    pid->landId = 0xFFFF;
+    pid->looks = 6;
+    mNPS_reset_schedule_area(pid);
+    mLd_ClearLandName(pid->landName);
+}
+
+s32 mNpc_CheckFreeAnimalPersonalID(AnmPersonalID_c* pid) {
+    s32 res = FALSE;
+
+    if ((pid->npcId == 0 && pid->nameId == 0xFF) || (ACTOR_FGNAME_GET_F000(pid->npcId) != FGNAME_F000_E)) {
+        res = TRUE;
+    }
+
+    return res;
+}
+
+void mNpc_CopyAnimalPersonalID(AnmPersonalID_c* dst, AnmPersonalID_c* src) {
+    if ((src != NULL) && (ACTOR_FGNAME_GET_F000(src->npcId) == FGNAME_F000_E)) {
+        dst->npcId = src->npcId;
+        dst->nameId = src->nameId;
+        dst->landId = src->landId;
+        dst->looks = src->looks;
+        mLd_CopyLandName(dst->landName, src->landName);
+    }
+}
+
+s32 mNpc_CheckCmpAnimalPersonalID(AnmPersonalID_c* id0, AnmPersonalID_c* id1) {
+    s32 res = FALSE;
+
+    if ((id0 != NULL) && (id1 != NULL) && (id0->npcId == id1->npcId) && (id0->nameId == id1->nameId) &&
+        (id0->landId == id1->landId)) {
+        if (mLd_CheckCmpLandName(id0->landName, id1->landName) == TRUE) {
+            res = TRUE;
+        }
+    }
+    return res;
+}
+
+s32 mNpc_GetAnimalNum() {
+    Animal_c* animal = common_data.animals;
+    s32 num = 0;
+    s32 i;
+
+    for (i = 0; i < ANIMAL_NUM_MAX; i++) {
+        if (mNpc_CheckFreeAnimalPersonalID(&animal->id) == FALSE) {
+            num++;
+        }
+
+        animal++;
+    }
+
+    return num;
+}
+
+s32 mNpc_CheckRemoveExp(Animal_c* animal) {
+    s32 res = (animal->removeInfo >> 13) & 7;
+
+    return res;
+}
+
+s32 mNpc_GetRemoveTime(Animal_c* animal) {
+    s32 res = animal->removeInfo & 0x1FFF;
+
+    return res;
+}
+
+void mNpc_AddRemoveTime(Animal_c* animal) {
+    if ((animal->removeInfo & 0x1FFF) < 0x1FFF) {
+        animal->removeInfo++;
+    }
+}
+
+void mNpc_SetRemoveExp(Animal_c* animal, u16 removeExp) {
+    animal->removeInfo = ((((animal->removeInfo >> 13) & 7) >> 13) | removeExp) << 13;
+}
+
+void mNpc_SetParentName(Animal_c* animal, PersonalID_c* parentId) {
+    if (parentId != NULL && mPr_NullCheckPersonalID(parentId) == FALSE &&
+        mPr_NullCheckPlayerName(animal->parentName) == TRUE) {
+        mPr_CopyPlayerName(animal->parentName, parentId->playerName);
+    }
+}
+
+void mNpc_SetParentNameAllAnimal() {
+    Animal_c* animal = common_data.animals;
+    PrivateInfo* priv = common_data.privateInfo;
+    s32 i;
+
+    if (priv != NULL) {
+        for (i = 0; i < ANIMAL_NUM_MAX; i++) {
+            mNpc_SetParentName(animal, &priv->playerId);
+            animal++;
+        }
+    }
+}
+
+void mNpc_ClearAnimalMail(Anmplmail_c* mail) {
+    mail->font = 0xFF;
+    mail->paperType = 0;
+    mail->present = 0;
+    mem_clear(mail->body, sizeof(mail->body), 0x20);
```

Comment:
I have always wondered if we should use `' '` for this kind of thing. What do you think?
```suggestion
    mem_clear(mail->body, sizeof(mail->body), ' ');
```

Reply:
Regarding to the conversation in discord, the game uses a custom encoding which isn't shift-jis
So I believe it's better to leave them by value at the moment